src/main/scala/sbtassembly/MergeStrategy.scala

```diff
@@ -1,148 +1,245 @@
 package sbtassembly
 
-import sbt._
-import java.io.{ FileOutputStream, File}
-import Assembly.{ sha1string, sha1content }
+import sbt.io.{ IO, Using }
+import sbt.util.Level
+import sbtassembly.Assembly._
+import sbtassembly.AssemblyUtils.AppendEofInputStream
+
+import java.io.{ BufferedReader, ByteArrayInputStream, InputStreamReader, SequenceInputStream }
+import java.nio.charset.Charset
+import java.nio.file.Paths
+import java.util.Collections
+import scala.collection.JavaConverters._
+import scala.reflect.io.Streamable
 
 /**
- * MergeStrategy is invoked if more than one source file is mapped to the 
- * same target path. Its arguments are the tempDir (which is deleted after
- * packaging) and the sequence of source files, and it shall return the
- * file to be included in the assembly (or throw an exception).
+ * Class for merging files that shares the same target path
  */
-abstract class MergeStrategy extends Function1[(File, String, Seq[File]), Either[String, Seq[(File, String)]]] {
+trait MergeStrategy {
```

Comment:
> @eed3si9n The method signature of `MergeStrategy.merge` also has changed:
> 
> `merge(conflicts: Vector[Dependency]): Either[String, Vector[JarEntry]]` whereas the old one just uses standard primitive and collection types.
> 
> This is exposed to the user because they can create custom merge strategies. Shall we use contraband for these to at least guarantee some level of backward compatibility?

Thanks for highlighting this point. I think extending `Function1` is cooler than making `def merge`. If this is mostly implementation detail, I feel like we can do some fp here.

If we provide some kind of function transformer as `MergeStrategy.apply(...)` that takes old function, and returns a new `MergeStrategy` that would be great.

Reply:
> Thanks for highlighting this point. I think extending `Function1` is cooler than making `def merge`. If this is mostly implementation detail, I feel like we can do some fp here.

Well, in my experience, when I create a custom MergeStrategy in 1.2.0, it is _not very clear_ to me what I need to do as it is a function.
The method signature 

```scala
Function1[(File, String, Seq[File]), Either[String, Seq[(File, String)]]]
```
is confusing since a function does not name its parameters, unless I look at one of the existing implementations, where I can learn that the first argument is the `tempDir` - but it still is not clear how to use it or if I even need it.
The other parameters maybe a little bit clearer once one looks at the implementation.

```scala
val first: MergeStrategy = new MergeStrategy {
    val name = "first"
    def apply(tempDir: File, path: String, files: Seq[File]): Either[String, Seq[(File, String)]] =
      Right(Seq(files.head -> path))
  }
```

With the changed signature, it is very clear that I am given a trait to implement and I have to override the `merge` and the `name` methods.

```scala
  case _ =>
    new MergeStrategy {
      override def name = "First"

      override def merge(conflicts: Vector[Assembly.Dependency]) =
        Right(Vector(JarEntry(conflicts.head.target, conflicts.head.stream)))
    }
```
That is my main motivation for changing this part of the code - to make it easier to understand for users.

> If we provide some kind of function transformer as `MergeStrategy.apply(...)` that takes old function, and returns a new `MergeStrategy` that would be great.

I cannot retrofit the **old** function to the new one, as `File` represents **extracted files** when we talk of jars. For project classes, this is easy to make but then we lose uniformity of the signature with libraries.

The new `Dependency` type instead carries a lazy input stream payload of `() => InputStream` that can refer to the **content of an entry inside an _unextracted_ jar** plus of course the `target` path. If people wanted to find out `sourceOfFiileForMerge` in the old implementation, they can extract the `module: Option[ModuleCoordinate]` for jars or the `source` for project classes.

So, if possible, I wanted to keep the change, and maybe to increase backward compatibility for 2.x.x, I could learn to use `contraband` instead of case classes.

But, IMHO, the old MergeStrategy contract/semantics is very hard to understand.

Reply:
I see that the old impl may have been confusing, but with `Vector[Dependency]` being the only input now for `merge` isn't the semantics of `Vector[Dependency] => Either[String, Vector[JarEntry]]` clearer now?
Instead of `new MergeStrategy` we can create `apply(name: String)(f: Vector[Dependency] => Either[String, Vector[JarEntry]])` on the companion object so a custom strategy could look like:

```scala
MergeStrategy(name = "First") { conflicts =>
  Right(Vector(JarEntry(conflicts.head.target, conflicts.head.stream)))
}
```

re: Contraband, I don't see a huge upside to using it here since Contraband is basically a case class replacement that's capable of keeping binary compatibility and persisting into JSON.

Reply:
> I see that the old impl may have been confusing, but with `Vector[Dependency]` being the only input now for `merge` isn't the semantics of `Vector[Dependency] => Either[String, Vector[JarEntry]]` clearer now? Instead of `new MergeStrategy` we can create `apply(name: String)(f: Vector[Dependency] => Either[String, Vector[JarEntry]])` on the companion object so a custom strategy could look like:
> 
> ```scala
> MergeStrategy(name = "First") { conflicts =>
>   Right(Vector(JarEntry(conflicts.head.target, conflicts.head.stream)))
> }
> ```
> 
> re: Contraband, I don't see a huge upside to using it here since Contraband is basically a case class replacement that's capable of keeping binary compatibility and persisting into JSON.

Ahh I see what you mean now. This is definitely better. I just got confused when you mentioned "old" because even the old implementation used `new MergeStrategy(...)` for built-in merge strats.

It makes sense now